File: .github/leetcode-runner/helpers/python/leetcode_helper.py
# LeetCode Helper for Python
# データ構造の定義とパース関数

from typing import List, Optional
import json
import re

# ========================================
# データ構造定義
# ========================================

class ListNode:
    """リンクリストのノード"""
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __repr__(self):
        return f"ListNode({self.val})"


class TreeNode:
    """二分木のノード"""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __repr__(self):
        return f"TreeNode({self.val})"


# ========================================
# パース関数
# ========================================

def parse_list(s: str) -> List:
    """文字列をリストにパース"""
    s = s.strip()
    if not s:
        return []
    try:
        return json.loads(s)
    except json.JSONDecodeError:
        # カンマ区切りとして処理
        return [x.strip() for x in s.split(',')]


def parse_int(s: str) -> int:
    """文字列を整数にパース"""
    return int(s.strip())


def parse_float(s: str) -> float:
    """文字列を浮動小数点にパース"""
    return float(s.strip())


def parse_string(s: str) -> str:
    """文字列をパース（クォートを除去）"""
    s = s.strip()
    if (s.startswith('"') and s.endswith('"')) or (s.startswith("'") and s.endswith("'")):
        return s[1:-1]
    return s


def parse_bool(s: str) -> bool:
    """文字列を真偽値にパース"""
    s = s.strip().lower()
    return s in ('true', '1', 'yes')


# ========================================
# リンクリスト操作
# ========================================

def list_to_linked_list(arr: List[int]) -> Optional[ListNode]:
    """配列をリンクリストに変換"""
    if not arr:
        return None
    head = ListNode(arr[0])
    current = head
    for val in arr[1:]:
        current.next = ListNode(val)
        current = current.next
    return head


def linked_list_to_list(head: Optional[ListNode]) -> List[int]:
    """リンクリストを配列に変換"""
    result = []
    current = head
    while current:
        result.append(current.val)
        current = current.next
    return result


def parse_linked_list(s: str) -> Optional[ListNode]:
    """文字列をリンクリストにパース"""
    arr = parse_list(s)
    return list_to_linked_list(arr)


# ========================================
# 二分木操作
# ========================================

def list_to_tree(arr: List) -> Optional[TreeNode]:
    """LeetCode形式の配列を二分木に変換"""
    if not arr or arr[0] is None:
        return None

    root = TreeNode(arr[0])
    queue = [root]
    i = 1

    while queue and i < len(arr):
        node = queue.pop(0)

        # 左の子
        if i < len(arr):
            if arr[i] is not None:
                node.left = TreeNode(arr[i])
                queue.append(node.left)
            i += 1

        # 右の子
        if i < len(arr):
            if arr[i] is not None:
                node.right = TreeNode(arr[i])
                queue.append(node.right)
            i += 1

    return root


def tree_to_list(root: Optional[TreeNode]) -> List:
    """二分木をLeetCode形式の配列に変換"""
    if not root:
        return []

    result = []
    queue = [root]

    while queue:
        node = queue.pop(0)
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)

    # 末尾のNoneを削除
    while result and result[-1] is None:
        result.pop()

    return result


def parse_tree(s: str) -> Optional[TreeNode]:
    """文字列を二分木にパース"""
    arr = parse_list(s)
    return list_to_tree(arr)


# ========================================
# 入力パース（汎用）
# ========================================

def parse_input(s: str, type_hint: str = 'auto'):
    """
    入力文字列をパース
    type_hint: 'list', 'int', 'float', 'string', 'bool', 'linked_list', 'tree', 'auto'
    """
    s = s.strip()

    if type_hint == 'list':
        return parse_list(s)
    elif type_hint == 'int':
        return parse_int(s)
    elif type_hint == 'float':
        return parse_float(s)
    elif type_hint == 'string':
        return parse_string(s)
    elif type_hint == 'bool':
        return parse_bool(s)
    elif type_hint == 'linked_list':
        return parse_linked_list(s)
    elif type_hint == 'tree':
        return parse_tree(s)
    else:
        # 自動判定
        if s.startswith('['):
            return parse_list(s)
        elif s.isdigit() or (s.startswith('-') and s[1:].isdigit()):
            return parse_int(s)
        elif s.lower() in ('true', 'false'):
            return parse_bool(s)
        else:
            return parse_string(s)


# ========================================
# 出力フォーマット
# ========================================

def format_output(value) -> str:
    """出力値を文字列にフォーマット"""
    if isinstance(value, ListNode):
        return json.dumps(linked_list_to_list(value))
    elif isinstance(value, TreeNode):
        return json.dumps(tree_to_list(value))
    elif isinstance(value, bool):
        return 'true' if value else 'false'
    elif isinstance(value, (list, tuple)):
        return json.dumps(value)
    elif value is None:
        return 'null'
    else:
        return str(value)

File: .github/leetcode-runner/helpers/javascript/leetcode_helper.js
// LeetCode Helper for JavaScript/TypeScript
// データ構造の定義とパース関数

// ========================================
// データ構造定義
// ========================================

class ListNode {
  constructor(val = 0, next = null) {
    this.val = val;
    this.next = next;
  }
}

class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

// ========================================
// パース関数
// ========================================

function parseList(s) {
  s = s.trim();
  if (!s) return [];
  try {
    return JSON.parse(s);
  } catch {
    return s.split(',').map(x => x.trim());
  }
}

function parseInt_(s) {
  return parseInt(s.trim(), 10);
}

function parseFloat_(s) {
  return parseFloat(s.trim());
}

function parseString(s) {
  s = s.trim();
  if ((s.startsWith('"') && s.endsWith('"')) || (s.startsWith("'") && s.endsWith("'"))) {
    return s.slice(1, -1);
  }
  return s;
}

function parseBool(s) {
  s = s.trim().toLowerCase();
  return s === 'true' || s === '1' || s === 'yes';
}

// ========================================
// リンクリスト操作
// ========================================

function listToLinkedList(arr) {
  if (!arr || arr.length === 0) return null;
  const head = new ListNode(arr[0]);
  let current = head;
  for (let i = 1; i < arr.length; i++) {
    current.next = new ListNode(arr[i]);
    current = current.next;
  }
  return head;
}

function linkedListToList(head) {
  const result = [];
  let current = head;
  while (current) {
    result.push(current.val);
    current = current.next;
  }
  return result;
}

function parseLinkedList(s) {
  const arr = parseList(s);
  return listToLinkedList(arr);
}

// ========================================
// 二分木操作
// ========================================

function listToTree(arr) {
  if (!arr || arr.length === 0 || arr[0] === null) return null;

  const root = new TreeNode(arr[0]);
  const queue = [root];
  let i = 1;

  while (queue.length > 0 && i < arr.length) {
    const node = queue.shift();

    // 左の子
    if (i < arr.length) {
      if (arr[i] !== null) {
        node.left = new TreeNode(arr[i]);
        queue.push(node.left);
      }
      i++;
    }

    // 右の子
    if (i < arr.length) {
      if (arr[i] !== null) {
        node.right = new TreeNode(arr[i]);
        queue.push(node.right);
      }
      i++;
    }
  }

  return root;
}

function treeToList(root) {
  if (!root) return [];

  const result = [];
  const queue = [root];

  while (queue.length > 0) {
    const node = queue.shift();
    if (node) {
      result.push(node.val);
      queue.push(node.left);
      queue.push(node.right);
    } else {
      result.push(null);
    }
  }

  // 末尾のnullを削除
  while (result.length > 0 && result[result.length - 1] === null) {
    result.pop();
  }

  return result;
}

function parseTree(s) {
  const arr = parseList(s);
  return listToTree(arr);
}

// ========================================
// 入力パース（汎用）
// ========================================

function parseInput(s, typeHint = 'auto') {
  s = s.trim();

  switch (typeHint) {
    case 'list':
      return parseList(s);
    case 'int':
      return parseInt_(s);
    case 'float':
      return parseFloat_(s);
    case 'string':
      return parseString(s);
    case 'bool':
      return parseBool(s);
    case 'linked_list':
      return parseLinkedList(s);
    case 'tree':
      return parseTree(s);
    default:
      // 自動判定
      if (s.startsWith('[')) {
        return parseList(s);
      } else if (/^-?\d+$/.test(s)) {
        return parseInt_(s);
      } else if (s.toLowerCase() === 'true' || s.toLowerCase() === 'false') {
        return parseBool(s);
      } else {
        return parseString(s);
      }
  }
}

// ========================================
// 出力フォーマット
// ========================================

function formatOutput(value) {
  if (value instanceof ListNode) {
    return JSON.stringify(linkedListToList(value));
  } else if (value instanceof TreeNode) {
    return JSON.stringify(treeToList(value));
  } else if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  } else if (Array.isArray(value)) {
    return JSON.stringify(value);
  } else if (value === null || value === undefined) {
    return 'null';
  } else {
    return String(value);
  }
}

// エクスポート
module.exports = {
  ListNode,
  TreeNode,
  parseList,
  parseInt: parseInt_,
  parseFloat: parseFloat_,
  parseString,
  parseBool,
  listToLinkedList,
  linkedListToList,
  parseLinkedList,
  listToTree,
  treeToList,
  parseTree,
  parseInput,
  formatOutput,
};
